### ISSUE TITLE
Marshalling Rust enums

### ISSUE THREAD

**Issue by Laegluin** [gluon-lang/gluon#539]
How can I marshal a Rust enum to Gluon? I only need an untagged enum, but I assume there is some way to marshal enums that are only made up of types that implement `VmType`. I know you can simply define the enum in Gluon, but that is really inconvenient when you it to be part of a native module, so that you can import from the same module as the functions that take it as argument.

If this is possible, it would probably not hurt to add a marshalling chapter in the book, explaining how to work with complex types. I'll try to do that as soon as I get it working :)

**Comment by Marwes** [comment]
Hello!

So there is a marshalling example at https://github.com/gluon-lang/gluon/blob/master/examples/marshalling.rs which I believe should cover what you want, it is slightly awkward in how you need to use the generated code but it should at least be automatic once you get past that.

Having a chapter covering the topc in the book would definitely be a good idea as well.

**Comment by Laegluin** [comment]
Thanks, that seems like I can make it work!

So as far as I can tell, I'll have to generate the source code for the enum and then load it as a module I can import from the actual module I want to expose. Does that mean I have to always write a Gluon wrapper for my module?

Lets say I want to add some module like this:

```rust
let module = record! {
    function => primitive!(2 some_function),
};
```

Ideally I could just add an enum that my function needs like this:

```rust
let module = record! {
    function => primitive!(2 some_function),
    Enum => some_macro!(Enum),
};
```

If I understand the example correctly, what I have to do instead is:

- generate the code for the enum and add it as module
- add the module containing the functions I want to add
- create a third module that imports both the enum and the functions and exposes them as a single module

It just seems like a lot of work for something that seems to be a quite common use case. Is it simply not possible to add a type using the `record!` macro?

**Comment by Marwes** [comment]
> It just seems like a lot of work for something that seems to be a quite common use case. Is it simply not possible to add a type using the record! macro?

Yeah, that is not very user friendly. It is possible to extend `record!` to support types as well though, I started to add it with https://github.com/gluon-lang/gluon/pull/540 as it will make it possible to export some types in the standard library as well (which currently cannot be refereed to from gluon scripts). Still need to verify that it works as I want on the http example though which needs to do this as well.

**Comment by Laegluin** [comment]
Looks great!

I'll try to make things works on my side, and then add some stuff to the docs. With the new `record` macro, would you still need to use serde for making enums work? Even if it's a lot of boilerplate otherwise, you could probably use a `derive` macro for that (I was thinking about doing something similar for implementing `Userdata`).

**Comment by Marwes** [comment]
> With the new record macro, would you still need to use serde for making enums work?

Yep, long term it might be worthwhile to add a dedicated trait + derive but it isn't anything I have planned yet. The serde integration is nice regardless though as types from external crates is more likely to have an implementation of `Serialize/Deserialize` than "GluonMarshal".

I don't think I have an issue open for a "GluonMarshal" trait so if you find some case which `Serialize/Deserialize` do not cover feel free to open an issue for it to gather some motivating usecases.

**Comment by Marwes** [comment]
The `Getable/Pushable` pair of traits is probably what should become a more general `GluonMarshal` pair of of traits.

**Comment by Laegluin** [comment]
> The serde integration is nice regardless though as types from external crates is more likely to have an implementation of Serialize/Deserialize than "GluonMarshal".

Yeah, that makes sense. I have played around with the api, and to me `Getable/Pushable` are really what's annoying to implement. Being able to automatically generate Gluon code for your types is really helpful, but most people would probably be fine with having to do that manually anyway.

Deriving `Getable` and `Pushable` seems to be fairly straight forward for most cases, I think I'll have a look and try to implement at least a basic macro for that.

**Comment by Laegluin** [comment]
Closed by #541.
